pkg/content/content.go

```diff
@@ -0,0 +1,185 @@
+		// TODO: support multi-platform source images.
+		containerd.WithImageHandler(images.HandlerFunc(
+			func(ctx context.Context, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+				if images.IsLayerType(desc.MediaType) {
+					logger.Debugf("pulling layer %s", desc.Digest)
+				}
+				return nil, nil
+			},
+		)),
+		containerd.WithPullUnpack,
+		containerd.WithResolver(resolver),
+	}
+
+	// Pull the source image from remote registry.
+	image, err := pvd.client.Pull(ctx, ref, opts...)
+	if err != nil {
+		return errors.Wrap(err, "pull source image")
+	}
+	pvd.image = image
+
+	return nil
+}
+
+func (pvd *LocalContentProvider) Write(ctx context.Context, desc ocispec.Descriptor, reader io.Reader) error {
+	digest := desc.Digest.String()
+
+	gcLabels := map[string]string{
+		"containerd.io/gc.ref.content": digest,
+	}
+	if err := content.WriteBlob(
+		ctx, pvd.client.ContentStore(), digest, reader, desc, content.WithLabels(gcLabels),
+	); err != nil {
+		return errors.Wrapf(err, "write blob %s", digest)
+	}
+
+	return nil
+}
+
+func (pvd *LocalContentProvider) Read(ctx context.Context, desc ocispec.Descriptor) ([]byte, error) {
+	return content.ReadBlob(ctx, pvd.client.ContentStore(), desc)
+}
+
+func (pvd *LocalContentProvider) Push(ctx context.Context, desc ocispec.Descriptor, ref string) error {
+	refUrl, err := url.Parse(fmt.Sprintf("dummy://%s", ref))
+	if err != nil {
+		return errors.Wrap(err, "parse reference")
+	}
+	remote, err := remote.NewRemote(ref, func() remotes.Resolver {
+		var resolver remotes.Resolver
+		if auth, ok := pvd.cfg.Source[refUrl.Host]; ok {
+			resolver = remote.NewResolver(auth.Insecure, remote.NewBasicAuthCredFunc(auth.Auth))
+		} else {
+			resolver = remote.NewResolver(false, remote.NewDockerConfigCredFunc())
+		}
+		return resolver
+	})
+	if err != nil {
+		return errors.Wrap(err, "create remote")
+	}
+
+	pushHandler := images.HandlerFunc(func(ctx context.Context, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+		pusher, err := remote.Pusher(ctx, ref)
+		if err != nil {
+			return nil, errors.Wrap(err, "create pusher")
+		}
+		return remotes.PushHandler(pusher, pvd.client.ContentStore()).Handle(ctx, desc)
+	})
+
+	// Get the manifest of target image from content store.
+	bytes, err := content.ReadBlob(ctx, pvd.client.ContentStore(), desc)
+	if err != nil {
+		return errors.Wrapf(err, "read from content store: %v", desc)
+	}
+	var manifest ocispec.Manifest
+	if err := json.Unmarshal(bytes, &manifest); err != nil {
+		return errors.Wrapf(err, "unmarshal target manifest")
+	}
+	descs := []ocispec.Descriptor{manifest.Config}
+	descs = append(descs, manifest.Layers...)
+
+	// Push the target image config and all layers to remote registry.
+	if err := images.Dispatch(ctx, pushHandler, nil, descs...); err != nil {
+		return errors.Wrapf(err, "push config and layer")
+	}
+
+	// Push the manifest of target image to remote registry.
+	return images.Dispatch(ctx, pushHandler, nil, desc)
```

Comment:
Why not use [Push API](https://github.com/containerd/containerd/blob/483e23bdcdf46ae204f267d6864d1a51da7ec1ae/client.go#L421)?